src/pages/experiments/[experimentId]/data-exploration/components/heatmap/HeatmapPlot.jsx

```diff
@@ -121,83 +122,148 @@ const HeatmapPlot = (props) => {
     }
   }, [width]);
 
-  const downsample = (groupBy) => {
+  // Defines sorting order according to which group the cell belongs to
+  const positionInGrouping = (cellSetNames, cell) => {
+    let position = 0;
+
+    for (const { key: cellSetName } of cellSetNames) {
+      if (cellsInSet(cellSetName).has(cell)) {
+        return position;
+      }
+
+      position += 1;
+    }
+
+    // If not in grouping set it in the end
+    return position + 1;
+  };
+
+  const cellsInSet = (cellSetName) => {
+    return properties[cellSetName].cellIds;
+  };
+
+  ///////////////////////////////////////////////////////////////////////////////////////////
+  // Returns a set of all cells relevant to the experiment (so the ones belonging to one or more of the enabled sets and not hidden).
+  const getAllRelevantCellIds = (arrayOfcellSetNamesByGroupByKey) => {
     // Find all hidden cells.
     const hiddenCellIds = union(Array.from(hidden), properties);
 
-    // Get how many cells we are sampling.
-    let total = 0;
+    let allCellsInSets = new Set();
 
-    // Create an object for storing the cells grouped by `groupBy`.
-    const groupedCells = {};
+    Object.values(arrayOfcellSetNamesByGroupByKey).forEach((arrayOfSetNames) => {
+      // Add each set of cells into 
+      arrayOfSetNames.forEach((setNameObj) => {
+        const cellSet = cellsInSet(setNameObj.key);
+        // Union of allCellsInSets and cellSet
+        allCellsInSets = new Set([...allCellsInSets, ...cellSet]);
+      });
+    });
 
-    // Find the `groupBy` root node.
-    const rootNode = hierarchy.filter((clusters) => clusters.key === groupBy);
+    // Only work with non-hidden cells.
+    const shownCells = new Set([...allCellsInSets].filter(x => !hiddenCellIds.has(x)));
 
-    if (!rootNode.length) {
-      return [];
+    return shownCells;
+  }
+
+  const getCellSetNamesDict = (rootNodes) => {
+    const cellSetNamesByGroupByKey = {};
+    rootNodes.forEach((currentRootNode) => cellSetNamesByGroupByKey[currentRootNode.key] = currentRootNode.children);
+
+    return cellSetNamesByGroupByKey
+  };
+
+  const downsampled = (cellIds) => {
```

Comment:
this seems like a very heavy rewrite - I wonder whether there isn't an easier way to implement this that reuses the downsample function that we already have in master, but leverage it to be used across all clusters, not per cluster

Reply:
I am not sure how we could use it across all groupBy's because we would have to keep in mind the intersections of cells between each group by, and if we went into each groupBy then we would run into many issues when trying to merge these different downsampled groups of cells into one downsample (like we can't guarantee that the merged downsampled set is actually less than maxCells so we would have to do a random downsampling after anyways).

This is a problem I ran into and asked about in the bioinformatics slack I believe, the random downsampling was what most people agreed with I believe.

Reply:
I fell like a solution trying to use the same logic of the original downsampling function would actually involve a heavy rewrite anyways beause we would have to introduce some kind of relationship between the cell groups to know which cells to pick to keep representativeness.

Reply:
Make sense, I understand your reasoning now and I agree that this is better.

Reply:
actually in the end I have made a big rewrite to this (marcell also made some comments about this in https://github.com/biomage-ltd/ui/pull/78#discussion_r563800792), so a lot of this is changed and it is more based on the original method of sampling.